### DISCUSSION TITLE
Crash while playing with `Assertion '!xcb_xlib_threads_sequence_lost' failed`

### DISCUSSION THREAD

**Discussion by nicolas-abril** [ppy/osu#21474] [Q&A] [answered]
### Type

Crash to desktop

### Bug description

Game started to crash while playing, out of nowhere, using a version that used to work before one or two weeks ago. I tried updating to the latest version of the game, but it still happens.

I'm running the AppImage version of osu! on Linux, using openSUSE Tumbleweed. The only change in my system compared to when the game used to work were the routine package updates that I do daily.

The logs don't seem to show anything about the crash. Launching the game through the terminal I get the following error message:
```
[xcb] Unknown sequence number while processing queue
[xcb] You called XInitThreads, this is not your fault
[xcb] Aborting, sorry about that.
osu!: xcb_io.c:278: poll_for_event: Assertion `!xcb_xlib_threads_sequence_lost' failed.
Aborted (core dumped)
```

The crash doesn't always happen at a specific point in a song, sometimes it takes longer to happen, sometimes shorter and seems to happen with any song. In the video I attached in this issue it happened especially fast, but it's usually after 1-2 minutes of gameplay.

### Screenshots or videos

Video showing the crash happening: https://youtu.be/EUa3qPM_1aM

### Version

2022.1123.0-lazer

### Logs

[database.log](https://github.com/ppy/osu/files/10127750/database.log)
[input.log](https://github.com/ppy/osu/files/10127751/input.log)
[network.log](https://github.com/ppy/osu/files/10127752/network.log)
[performance.log](https://github.com/ppy/osu/files/10127753/performance.log)
[runtime.log](https://github.com/ppy/osu/files/10127754/runtime.log)

**Comment by bdach** [comment, marked as answer]
This is likely not an issue for this repository. We don't directly interact with X.

It's either an issue with one of the updated packages on your system or something to report to SDL2. I can't tell which as I can't reproduce, you'll have to do the due diligence on your own.

**Reply by nicolas-abril** [reply]
Thank you for your help, looks like it is indeed the latest X update (this week) that started this problem for me. I'll try to get help from the openSUSE people.

**Comment by easyteacher** [comment]
osu! crashed often recently, and I saw the same error.

```
[xcb] Unknown sequence number while processing queue
[xcb] You called XInitThreads, this is not your fault
[xcb] Aborting, sorry about that.
osu!: xcb_io.c:278: poll_for_event: Assertion `!xcb_xlib_threads_sequence_lost' failed.
Aborted (core dumped)
```

**Reply by easyteacher** [reply]
The backtrace shows buggy libSDL2 library in osu!, and recently libsdl2 was updated to fix a crash: https://build.opensuse.org/request/show/1039395

```
Module /tmp/.mount_osu-laTRCLSy/usr/bin/libbassmix.so without build-id.
                Module /tmp/.mount_osu-laTRCLSy/usr/bin/libbass_fx.so without build-id.
                Module /tmp/.mount_osu-laTRCLSy/usr/bin/libbass_fx.so
                Module /tmp/.mount_osu-laTRCLSy/usr/bin/libbass.so without build-id.
                Module /tmp/.mount_osu-laTRCLSy/usr/bin/libbass.so
                Stack trace of thread 13052:
                #0  0x00007fc0dde957ec __pthread_kill_implementation (libc.so.6 + 0x907ec)
                #1  0x00007fc0dde42816 __GI_raise (libc.so.6 + 0x3d816)
                #2  0x00007fc0dde2b8a6 __GI_abort (libc.so.6 + 0x268a6)
                #3  0x00007fc0dde2b72b __assert_fail_base (libc.so.6 + 0x2672b)
                #4  0x00007fc0dde3ac46 __GI___assert_fail (libc.so.6 + 0x35c46)
                #5  0x00007fc0284ff824 poll_for_event (libX11.so.6 + 0x45824)
                #6  0x00007fc0284ff8e6 poll_for_response (libX11.so.6 + 0x458e6)
                #7  0x00007fc02850334a _XEventsQueued (libX11.so.6 + 0x4934a)
                #8  0x00007fc0284dff04 XCheckIfEvent (libX11.so.6 + 0x25f04)
                #9  0x00007fc0207474fe n/a (/tmp/.mount_osu-laTRCLSy/usr/bin/libSDL2.so + 0x1474fe)
                ELF object binary architecture: AMD x86-64
```

**Reply by bdach** [reply]
I'm not sure how you got to this conclusion. The reporter above claims that the issue was at the X side rather than on the SDL side?

**Reply by easyteacher** [reply]
I extracted the appimage, replaced libSDL2.so with the updated library, and did not see a crash. Before that osu! was basically unplayable.

**Reply by bdach** [reply]
Zero trace about this in upstream, just a [random commit without explanation](https://github.com/libsdl-org/SDL/commit/d87048fd5a7dff55c4cb205dd29e527c9cc51225)... Questionable but okay.

@peppy may be good to get new SDL2 libs built to resolve this at some point. Probably not horribly urgent given linux userbase though - this looks to be a distro-specific issue too because I'm not seeing this on ubuntu.

**Reply by peppy** [reply]
If we go with tracking `master` it's going to require some potentially major updates (SDL became 3).

https://github.com/ppy/SDL2-CS/actions/runs/3669685437/jobs/6203687603#step:12:8

Might be best to find the last 2.x commit for now? If that has the fix in it.

**Reply by bdach** [reply]
The fix has been backported to the [`release-2.26.x`](https://github.com/libsdl-org/SDL/commits/release-2.26.x) branch, so should be fine to use that.

[`SDL2`](https://github.com/libsdl-org/SDL/commits/SDL2) should be also fine while being a bit more bleeding-edge. In theory.

**Reply by nicolas-abril** [reply]
> I'm not sure how you got to this conclusion. The reporter above claims that the issue was at the X side rather than on the SDL side?

For me it stopped after they reverted the libX11 and libX11-xcb package. The libsdl package didn't change

**Reply by bdach** [reply]
SDL has been bumped in https://github.com/ppy/osu-framework/pull/5579. It'll trickle down to game eventually.

**Reply by easyteacher** [reply]
Thank you!

**Comment by huupoke** [comment]
Actually, this whole dumpster fire started after `libx11` start enforcing thread safety by default (since `1.8`). Misbehaving clients get deadlocked, so people send patches to fix that, but then it cause more problems. See the bottom of the post for detail.
So if you want the previous behaviour for every app (including old, misbehaving apps), you should downgrade to as low as `1.7.5`.

<hr>

Also, #21239 seems to be the same issue (and someone has also posted the [same workaround](https://github.com/ppy/osu/discussions/21239#discussioncomment-4178807)).
The issue reported on the libx11 repository: https://gitlab.freedesktop.org/xorg/lib/libx11/-/issues/170

<hr>
Problematic commit timeline:

1. [Enforcing thread safety](https://gitlab.freedesktop.org/xorg/lib/libx11/-/commit/afcdb6fb0045c6186aa83d9298f327a7ec1b2cb9)
1. [Fix misbehaving clients](https://gitlab.freedesktop.org/xorg/lib/libx11/-/commit/79775575418fd6f8ee1c5e5bbe403df4606fb5b6)
1. [Fix the "Fix misbehaving clients" commit](https://gitlab.freedesktop.org/xorg/lib/libx11/-/commit/a9e845809bcaae22496bc8aa3ca252b410d5f39b)
1. [Revert the "Fix the "Fix misbehaving clients" commit"](https://gitlab.freedesktop.org/xorg/lib/libx11/-/commit/88399e01be679bfcc9a5e8922ffe2c47f0e56dee)

**Reply by Battle420** [reply]
`libx11` with version `1.8.2-2` makes the crash not happen for me.

**Reply by Cyliann** [reply]
Thanks for the workaround. For me it started crashing only after update to `1.8.3` and downgrading to `1.8.2` fixed it.

**Reply by easyteacher** [reply]
see also https://github.com/ValveSoftware/Dota-2/issues/2218

**Reply by Battle420** [reply]
`libx11` version `1.8.3-3` came out, and it no longer crashes.
It's fixed I guess?

**Comment by Cyliann** [comment]
It seems to be a libx11 issue and not osu.

See: https://github.com/ValveSoftware/Dota-2/issues/2218
and
https://gitlab.freedesktop.org/xorg/lib/libx11/-/merge_requests/176
